compiler/noirc_frontend/src/hir/mod.rs

```diff
@@ -37,15 +41,43 @@ pub enum FunctionNameMatch<'a> {
 
 impl Context {
     pub fn new(file_manager: FileManager, crate_graph: CrateGraph) -> Context {
-        Context {
+        let mut context = Context {
             def_interner: NodeInterner::default(),
             def_maps: BTreeMap::new(),
             visited_files: BTreeMap::new(),
             crate_graph,
             file_manager,
+            file_to_ast_cache: HashMap::default(),
+        };
+
+        context.populate_cache_with_file_manager();
+
+        context
+    }
+
+    // TODO: This method should live at a higher level
+    fn populate_cache_with_file_manager(&mut self) {
+        for file_id in self.file_manager.as_file_map().all_file_ids() {
+            let file_path = self.file_manager.path(*file_id);
+            let file_extension =
+                file_path.extension().expect("expected all file paths to have an extension");
+            // TODO: Another reason we may not want to have this method here
+            // TODO: is the fact that we want to not have the compiler worry
+            // TODO about the nr file extension
+            if file_extension != "nr" {
+                continue;
+            }
+            self.file_to_ast_cache.insert(*file_id, parse_file(&self.file_manager, *file_id));
         }
     }
```

Comment:
This is the main change, notice that with this we can now parallelize parsing and given that the FileIds are deterministic, we can also avoid parsing unchanged things we've already parsed.

On the other hand, it may be easier to always parse everything. A future problem is that incremental compilation may not work effectively if the FileIds are not deterministic -- we can just hash the file_path in order to get the file_id in that case.